File: plugins/probe-rat-layer/probe_layer/core/keepout.py
"""禁布区加载、膨胀、线段相交、咽喉检测（基于 shapely）。"""
from __future__ import annotations

from ..model import Point, RectZone, CircleZone, KeepoutZone, Wire
from ..config import LayeringConfig
from . import geometry as geo


def load_keepouts(specs: list[dict]) -> tuple[KeepoutZone, ...]:
    zones: list[KeepoutZone] = []
    for s in specs:
        shape = s.get("shape")
        if shape == "rect":
            zones.append(RectZone(s["zone_id"], float(s["xmin"]), float(s["ymin"]),
                                  float(s["xmax"]), float(s["ymax"])))
        elif shape == "circle":
            c = s["center"]
            zones.append(CircleZone(s["zone_id"], Point(float(c["x"]), float(c["y"])),
                                    float(s["radius"])))
        else:
            raise ValueError(f"未知禁布区形状: {shape!r}")
    return tuple(zones)


def zones_crossed_by(w: Wire, zones: tuple[KeepoutZone, ...],
                     margin: float = 0.0) -> tuple[str, ...]:
    line = geo.wire_line(w)
    return tuple(z.zone_id for z in zones
                 if line.intersects(geo.zone_geom(z).buffer(margin) if margin > 0
                                    else geo.zone_geom(z)))


def both_cross_same_zone(wa: Wire, wb: Wire,
                         zones: tuple[KeepoutZone, ...]) -> tuple[str, ...]:
    a = set(zones_crossed_by(wa, zones))
    b = set(zones_crossed_by(wb, zones))
    return tuple(sorted(a & b))


def in_zone_overlap_length(wa: Wire, wb: Wire, zone: KeepoutZone) -> float:
    return geo.in_zone_overlap_length(wa, wb, zone)


def pinch_zones(wires: tuple[Wire, ...], zones: tuple[KeepoutZone, ...],
                cfg: LayeringConfig) -> tuple[str, ...]:
    """咽喉检测：多条线挤过同一禁布区且区窄（不足以并排通过）→ 咽喉。

    bottleneck 近似 = 穿越线所需通道宽 / 区窄边宽，> 1 视为咽喉（占位启发式，待反标定）。
    """
    out: list[str] = []
    for z in zones:
        crossing = [w for w in wires if geo.wire_line(w).intersects(geo.zone_geom(z))]
        if len(crossing) < 2:
            continue
        if isinstance(z, RectZone):
            narrow = min(z.xmax - z.xmin, z.ymax - z.ymin)
        else:
            narrow = 2.0 * z.radius
        needed = sum(w.width + w.clearance for w in crossing)
        if narrow > 0 and needed / narrow > 1.0:
            out.append(z.zone_id)
    return tuple(out)

File: core-plugins/probe-rat-layer/src/keepout.rs
//! 禁布区加载、膨胀、线段相交、咽喉检测（移植自 Python `core/keepout.py`）。

use crate::config::LayeringConfig;
use crate::geometry;
use crate::model::{CircleZone, KeepoutZone, Point, RectZone, Wire};
use serde_json::Value;

pub fn load_keepouts(specs: &[Value]) -> Result<Vec<KeepoutZone>, String> {
    let mut zones = Vec::new();
    for s in specs {
        let shape = s.get("shape").and_then(|v| v.as_str()).unwrap_or("");
        if shape == "rect" {
            let z = RectZone {
                zone_id: s["zone_id"].as_str().unwrap_or("").to_string(),
                xmin: s["xmin"].as_f64().unwrap_or(0.0),
                ymin: s["ymin"].as_f64().unwrap_or(0.0),
                xmax: s["xmax"].as_f64().unwrap_or(0.0),
                ymax: s["ymax"].as_f64().unwrap_or(0.0),
            };
            zones.push(KeepoutZone::Rect(z));
        } else if shape == "circle" {
            let c = s.get("center").cloned().unwrap_or(Value::Null);
            let center = Point::new(
                c.get("x").and_then(|v| v.as_f64()).unwrap_or(0.0),
                c.get("y").and_then(|v| v.as_f64()).unwrap_or(0.0),
            );
            zones.push(KeepoutZone::Circle(CircleZone {
                zone_id: s["zone_id"].as_str().unwrap_or("").to_string(),
                center,
                radius: s["radius"].as_f64().unwrap_or(0.0),
            }));
        } else {
            return Err(format!("未知禁布区形状: {shape}"));
        }
    }
    Ok(zones)
}

pub fn zones_crossed_by(w: &Wire, zones: &[KeepoutZone], margin: f64) -> Vec<String> {
    let mut out = Vec::new();
    for z in zones {
        let crosses = match z {
            KeepoutZone::Rect(r) => {
                let m = if margin > 0.0 {
                    geometry::seg_in_rect_length(
                        w.start,
                        w.end,
                        &RectZone {
                            zone_id: r.zone_id.clone(),
                            xmin: r.xmin - margin,
                            ymin: r.ymin - margin,
                            xmax: r.xmax + margin,
                            ymax: r.ymax + margin,
                        },
                    )
                } else {
                    geometry::seg_in_rect_length(w.start, w.end, r)
                };
                m > geometry::EPS
            }
            KeepoutZone::Circle(c) => {
                let radius = c.radius + margin;
                geometry::seg_in_circle_length(
                    w.start,
                    w.end,
                    &CircleZone {
                        zone_id: c.zone_id.clone(),
                        center: c.center,
                        radius,
                    },
                ) > geometry::EPS
            }
        };
        if crosses {
            out.push(z.zone_id().to_string());
        }
    }
    out
}

pub fn both_cross_same_zone(wa: &Wire, wb: &Wire, zones: &[KeepoutZone]) -> Vec<String> {
    let a: crate::collections::HashSet<String> =
        zones_crossed_by(wa, zones, 0.0).into_iter().collect();
    let b: crate::collections::HashSet<String> =
        zones_crossed_by(wb, zones, 0.0).into_iter().collect();
    let mut out: Vec<String> = a.intersection(&b).cloned().collect();
    out.sort();
    out
}

pub fn in_zone_overlap_length(wa: &Wire, wb: &Wire, zone: &KeepoutZone) -> f64 {
    geometry::in_zone_overlap_length(wa, wb, zone)
}

/// 咽喉检测：多条线挤过同一禁布区且区窄（不足以并排通过）→ 咽喉。
pub fn pinch_zones(wires: &[Wire], zones: &[KeepoutZone], _cfg: &LayeringConfig) -> Vec<String> {
    let mut out = Vec::new();
    for z in zones {
        let crossing: Vec<&Wire> = wires
            .iter()
            .filter(|w| zones_crossed_by(w, std::slice::from_ref(z), 0.0).len() > 0)
            .collect();
        if crossing.len() < 2 {
            continue;
        }
        let narrow = match z {
            KeepoutZone::Rect(r) => (r.xmax - r.xmin).min(r.ymax - r.ymin),
            KeepoutZone::Circle(c) => 2.0 * c.radius,
        };
        let needed: f64 = crossing.iter().map(|w| w.width + w.clearance).sum();
        if narrow > 0.0 && needed / narrow > 1.0 {
            out.push(z.zone_id().to_string());
        }
    }
    out
}
